core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -258,11 +258,14 @@ abstract class RDD[T: ClassTag](
    * subclasses of RDD.
    */
   final def iterator(split: Partition, context: TaskContext): Iterator[T] = {
-    if (storageLevel != StorageLevel.NONE) {
+    val iter = if (storageLevel != StorageLevel.NONE) {
       SparkEnv.get.cacheManager.getOrCompute(this, split, context, storageLevel)
     } else {
       computeOrReadCheckpoint(split, context)
     }
+    checkpointData.collect { case data: ReliableRDDCheckpointData[T] =>
+      data.getCheckpointIterator(this, iter, context, split.index)
```

Comment:
if you make `getCheckpointIterator` a method in the abstract `RDDCheckpointData` then you can just do
```
checkpointData
  .map(_.getCheckpointIterator(this, iter, context, split.index))
  .getOrElse(iter)
```
We can address local checkpointing separately and for now just have it return the iterator
```
def getCheckpointIterator(...): Iterator[T] = iter
```